crates/sui-core/src/authority_active/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,252 @@
+            metrics,
+        })
+    }
+
+    pub async fn run(mut self) {
+        let mut finished: FuturesOrdered<JoinHandle<(u64, VerifiedCheckpoint)>> =
+            FuturesOrdered::new();
+        let task_limit = TASKS_PER_CORE * num_cpus::get();
+
+        // In memory only "watermark" to keep track of the highest checkpoint for which we've
+        // spun up an execution task. This is to prevent us from rescheduling the same checkpoints
+        // in the case where we receive a synced checkpoint message before scheduled checkpoints have
+        // completed, and hence have had the chance to update the on-disk watermark `highest_executed_checkpoint`.
+        let mut highest_scheduled_checkpoint: Option<u64> = None;
+
+        loop {
+            tokio::select! {
+                // Check for completed workers and ratchet the highest_checkpoint_executed
+                // watermark accordingly. Note that given that checkpoints are guaranteed to
+                // be processed (added to FuturesOrdered) in seq_number order, using FuturesOrdered
+                // guarantees that we will also ratchet the watermarks in order.
+                Some(Ok((seq_num, checkpoint))) = finished.next() => {
+                    self.highest_executed_checkpoint = Some(seq_num as CheckpointSequenceNumber);
+                    self.metrics.last_executed_checkpoint.set(seq_num as i64);
+                    self.update_lag_metric(seq_num)
+                        .unwrap_or_else(|err| error!("Update lag metric error: {:?}", err));
+
+                    self.checkpoint_store
+                        .update_highest_executed_checkpoint(&checkpoint)
+                        .unwrap();
+                },
+                // Limits to `task_limit` worker tasks. If we are at capacity, reloop and either
+                // check again or attempt to process and consume the FuturesOrdered queue for
+                // finished tasks
+                received = self.mailbox.recv(), if finished.len() < task_limit => match received {
+                    Ok(checkpoint) => {
+                        let last_to_exec = checkpoint.sequence_number() as u64;
+
+                        // Note that either of these can be higher. If the node crashes with many
+                        // scheduled tasks, then the in-memory watermark starts as None, but the
+                        // persistent watermark is set, hence we start there. If we get a new
+                        // messsage with checkpoints tasks scheduled, then the in-memory watermark
+                        // will be greater, and hence we start from there.
+                        let next_to_exec = std::cmp::max(
+                            self.highest_executed_checkpoint
+                                .map(|highest| highest + 1)
+                                .unwrap_or(0),
+                            highest_scheduled_checkpoint
+                                .map(|highest| highest + 1)
```

Comment:
nit: i would probably change `highest + 1` to `highest.saturating_add(1)` in order to be explicit about overflow behavior. These are u64s so it shouldn't be an issue, but i like to be explicit when possible.